### PR TITLE
Fix c2d address detection in download endpoint

### LINKED ISSUE #418: Wrong c2d address detection
On the download Endpoint, we have a piece of code which detects if the files is being downloaded by C2d.  If yes, will allow that download, even is the service is "compute" type:
https://github.com/oceanprotocol/provider/blob/v4main/ocean_provider/routes/consume.py#L276-L287

This check doesn't take into account the computeEnvs, so for sure, it will not work

### PR DESCRIPTION
**Pull request by md0x**
Signed-off-by: Pablo Maldonado <pablo@oceanprotocol.com>

Fixes #418  .

Changes proposed in this PR:

- Retrieve c2d environments and compare consumer address with the service id to allow or block download

### CODE REVIEW

File: ocean_provider/routes/consume.py [hunk 1]
```diff
@@ -273,15 +273,20 @@ def download():
...
-        get_compute_info() if os.getenv("OPERATOR_SERVICE_URL") else None,
-        None,
+    # allow our C2D to download a compute asset
+    c2d_environments = get_c2d_environments()
+
+    is_c2d_consumer_address = bool(
+        [
+            True
+            for env in c2d_environments
+            if service.id == env["id"]
```
Comment (alexcos20) [inline ocean_provider/routes/consume.py:283]: why do have service.id here ?
we just need to compare addresses..

**Review by alexcos20 — CHANGES_REQUESTED** [review]

File: tests/test_download.py [hunk 2]
```diff
@@ -220,3 +222,117 @@ def test_download_multiple_files(client, publisher_wallet, consumer_wallet, web3
...
+        service.index,
+        get_provider_fees(asset.did, service, consumer_wallet.address, 0),
+        consumer_wallet,
+    )
+
+    nonce = str(datetime.utcnow().timestamp())
+    _msg = f"{asset.did}{nonce}"
+
+    # Consume using url index and auth token
+    # (let the provider do the decryption)
```
Comment (DMats) [inline tests/test_download.py:249]: I think this comment is incorrect. There's no auth token here. And when you say `(let the provider do the decryption)` I think you mean, `(let the provider fetch on-chain ddo using tx_id)`

The provider always does the decryption, whether it fetches the ddo from chain, or its provided in args.

File: tests/test_download.py [hunk 3]
```diff
@@ -220,3 +222,117 @@ def test_download_multiple_files(client, publisher_wallet, consumer_wallet, web3
...
+        "documentId": asset.did,
+        "serviceId": service.id,
+        "consumerAddress": consumer_wallet.address,
+        "signature": sign_message(_msg, consumer_wallet),
+        "transferTxId": tx_id,
+        "fileIndex": 0,
+        "nonce": nonce,
+    }
+
+    def other_service(*args, **kwargs):
```
Comment (DMats) [inline tests/test_download.py:260]: `*args, **kwargs` are unused.

File: tests/test_download.py [hunk 4]
```diff
@@ -220,3 +222,117 @@ def test_download_multiple_files(client, publisher_wallet, consumer_wallet, web3
...
+        new_service = copy.deepcopy(service)
+        new_service.__setattr__("type", "compute")
```
Comment (DMats) [inline tests/test_download.py:262]: I think this could be more readable as `new_service.type = "compute"`.

File: tests/test_download.py (hunk 2 above)
Comment (DMats) [reply tests/test_download.py:249]: Or you could just remove the comment.

File: tests/test_download.py [hunk 5]
```diff
@@ -220,3 +222,117 @@ def test_download_multiple_files(client, publisher_wallet, consumer_wallet, web3
...
+        return new_service
```
Comment (DMats) [inline tests/test_download.py:319]: See above comments.

File: tests/test_download.py [hunk 6]
```diff
@@ -220,3 +222,117 @@ def test_download_multiple_files(client, publisher_wallet, consumer_wallet, web3
...
+
+    with patch("ocean_provider.routes.consume.get_c2d_environments") as mock:
+        mock.return_value = [
+            {
+                "priceMin": 60,
```
Comment (DMats) [inline tests/test_download.py:268]: Why is `priceMin` needed here?

**Review by DMats — CHANGES_REQUESTED** [review]

**Review by DMats — APPROVED** [review]

**Review by alexcos20 — APPROVED** [review]

**Pull request merged**